File: ios/RNWebGLTextureLoader.m
#import "RNWebGLTextureLoader.h"
#import "RNWebGLTexture.h"
#import <React/RCTLog.h>
#import <React/RCTUtils.h>


@implementation RNWebGLTextureLoader {
  NSArray<id<RNWebGLTextureConfigLoader>> *_loaders;
  NSMutableDictionary<NSNumber *, RNWebGLTexture *> *_objects;
}

@synthesize bridge = _bridge;

RCT_EXPORT_MODULE()

- (instancetype)init
{
  if ((self = [super init])) {
    _objects = [NSMutableDictionary dictionary];
  }
  return self;
}

-(id<RNWebGLTextureConfigLoader>) objectLoaderForConfig:(NSDictionary *)config {
  if (!_loaders) {
    _loaders = [_bridge modulesConformingToProtocol:@protocol(RNWebGLTextureConfigLoader)];
  }
  for (id<RNWebGLTextureConfigLoader> loader in _loaders) {
    if ([loader canLoadConfig:config]) {
      return loader;
    }
  }
  return nil;
}

-(void)loadWithConfig:(NSDictionary *)config
  withCompletionBlock:(RNWebGLTextureCompletionBlock)callback {
  id<RNWebGLTextureConfigLoader> loader = [self objectLoaderForConfig:config];
  if (!loader) {
    if (RCT_DEBUG) RCTLogError(@"No suitable RNWebGLTextureLoader found for %@", config);
    callback([NSError errorWithDomain:@"RNWebGL" code:1 userInfo:@{ NSLocalizedDescriptionKey: @"No suitable RNWebGLTextureLoader found" }], nil);
  }
  else {
    __weak RNWebGLTextureLoader *weakSelf = self;
    [loader loadWithConfig:config withCompletionBlock:^(NSError *err, RNWebGLTexture *obj) {
      if (obj && weakSelf) {
        weakSelf.objects[@(obj.objId)] = obj;
      }
      callback(err, obj);
    }];
  }
}

-(void)loadWithConfigAndWaitAttached:(NSDictionary *)config
                 withCompletionBlock:(RNWebGLTextureCompletionBlock)callback {
  [self loadWithConfig:config withCompletionBlock:^(NSError *error, RNWebGLTexture *obj) {
    if ([obj isAttached]) {
      callback(error, obj);
    }
    else {
      [obj listenAttached:^{
        callback(error, obj);
      }];
    }
  }];
}

-(void)unloadWithObjId:(RNWebGLTextureId)objId {
  NSNumber *key = @(objId);
  RNWebGLTexture *t = _objects[key];
  if (t) {
    [t unload];
    [_objects removeObjectForKey:key];
  }
}

-(void)unloadWithCtxId:(RNWebGLContextId)ctxId {
  NSMutableArray *unloadedKeys = [NSMutableArray array];
  for (NSNumber *key in [_objects keyEnumerator]) {
    RNWebGLTexture *t = _objects[key];
    if (t.ctxId == ctxId) {
      [t unload];
      [unloadedKeys addObject:key];
    }
  }
  [_objects removeObjectsForKeys:unloadedKeys];
}

@end

@implementation RCTBridge (RNWebGLTextureLoader)

- (RNWebGLTextureLoader *)webglObjectLoader
{
  return [self moduleForClass:[RNWebGLTextureLoader class]];
}

@end

File: android/src/main/java/fr/greweb/rnwebgl/RNWebGLTextureLoader.java
package fr.greweb.rnwebgl;

import android.util.Log;
import android.util.SparseArray;

import com.facebook.react.bridge.NativeModule;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReadableMap;

import java.util.ArrayList;
import java.util.List;

public class RNWebGLTextureLoader extends ReactContextBaseJavaModule {
    private List<RNWebGLTextureConfigLoader> mLoaders = null;
    private SparseArray<RNWebGLTexture> mObjects = new SparseArray<>();

    public RNWebGLTextureLoader(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @Override
    public String getName() {
        return "RNWebGLTextureLoader";
    }

    public RNWebGLTextureConfigLoader objectLoaderForConfig (final ReadableMap config) {
        if (mLoaders == null) {
            mLoaders = new ArrayList<>();
            for (NativeModule module: this.getReactApplicationContext().getCatalystInstance().getNativeModules()) {
                if (module instanceof RNWebGLTextureConfigLoader) {
                    mLoaders.add((RNWebGLTextureConfigLoader) module);
                }
            }
        }
        for (RNWebGLTextureConfigLoader loader : mLoaders) {
            if (loader.canLoadConfig(config)) {
                return loader;
            }
        }
        return null;
    }

    public void loadWithConfig (final ReadableMap config, final RNWebGLTextureCompletionBlock callback) {
        RNWebGLTextureConfigLoader loader = this.objectLoaderForConfig(config);
        if (loader == null) {
            Log.e("RNWebGL", "No suitable RNWebGLTextureLoader found for " + config);
            callback.call(new TextureLoaderNotFoundException(), null);
        }
        else {
            final RNWebGLTextureLoader self = this;
            loader.loadWithConfig(config, new RNWebGLTextureCompletionBlock() {
                public void call (Exception e, RNWebGLTexture obj) {
                    if (obj != null) {
                        self.mObjects.put(obj.objId, obj);
                    }
                    callback.call(e, obj);
                }
            });
        }
    }

    public void loadWithConfigAndWaitAttached (final ReadableMap config, final RNWebGLTextureCompletionBlock callback) {
        loadWithConfig(config, new RNWebGLTextureCompletionBlock() {
            @Override
            public void call(final Exception e, final RNWebGLTexture obj) {
                obj.listenAttached(new Runnable() {
                    public void run() {
                        callback.call(e, obj);
                    }
                });
            }
        });
    }

    public void unloadWithObjId (int objId) {
        RNWebGLTexture obj = mObjects.get(objId);
        if (obj != null) {
            mObjects.remove(objId);
            obj.destroy();
        }
    }

    public void unloadWithCtxId (int ctxId) {
        SparseArray<RNWebGLTexture> remaining = new SparseArray<>();
        for(int i = 0; i < mObjects.size(); i++) {
            int objId = mObjects.keyAt(i);
            RNWebGLTexture obj = mObjects.get(objId);
            if (obj.ctxId == ctxId) {
                obj.destroy();
            }
            else {
                remaining.put(objId, obj);
            }
        }
        mObjects = remaining;
    }
}
